File: python/lexer.py
"""CCHQCode Runtime - 词法分析器"""

import re
from typing import List


class Token:
    def __init__(self, type_: str, value: str, line: int = 0, col: int = 0):
        self.type = type_
        self.value = value
        self.line = line
        self.col = col

    def __repr__(self):
        return f'Token({self.type}, {self.value!r})'


KEYWORDS = {
    'if', 'else', 'for', 'while', 'break', 'continue',
    'switch', 'case', 'default', 'return', 'true', 'false'
}

SYMBOLS_SINGLE = '{}();,:[]=+-*/%!<>'
SYMBOL_MAP = {
    '==': '==', '!=': '!=', '<=': '<=', '>=': '>=',
    '&&': '&&', '||': '||', '=>': '=>', '->': '->',
}


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.length = len(source)
        self.pos = 0
        self.line = 1
        self.col = 1

    def peek(self, offset: int = 0) -> str:
        p = self.pos + offset
        return self.source[p] if p < self.length else ''

    def tokenize(self) -> List[Token]:
        tokens = []
        while self.pos < self.length:
            ch = self.source[self.pos]

            # 空白
            if ch.isspace():
                if ch == '\n':
                    self.line += 1
                    self.col = 1
                else:
                    self.col += 1
                self.pos += 1
                continue

            # 注释 //
            if ch == '/' and self.peek(1) == '/':
                self.pos += 2
                while self.pos < self.length and self.source[self.pos] != '\n':
                    self.pos += 1
                continue

            # @ 控件
            if ch == '@':
                self.pos += 1
                name = ''
                while self.pos < self.length and re.match(r'[a-zA-Z0-9_:]', self.source[self.pos]):
                    name += self.source[self.pos]
                    self.pos += 1
                tokens.append(Token('CONTROL', name, self.line, self.col))
                self.col += len(name) + 1
                continue

            # 变量 $xxx
            if ch == '$':
                self.pos += 1
                name = ''
                while self.pos < self.length and re.match(r'[a-zA-Z0-9_]', self.source[self.pos]):
                    name += self.source[self.pos]
                    self.pos += 1
                tokens.append(Token('VARIABLE', name, self.line, self.col))
                self.col += len(name) + 1
                continue

            # 字符串
            if ch in ('"', "'"):
                quote = ch
                self.pos += 1
                s = ''
                while self.pos < self.length and self.source[self.pos] != quote:
                    if self.source[self.pos] == '\\':
                        self.pos += 1
                        s += self.source[self.pos] if self.pos < self.length else ''
                    else:
                        s += self.source[self.pos]
                    self.pos += 1
                self.pos += 1  # 跳过结束引号
                tokens.append(Token('STRING', s, self.line, self.col))
                self.col += len(s) + 2
                continue

            # 数字
            if ch.isdigit() or (ch == '-' and self.peek(1).isdigit()):
                num = ''
                if ch == '-':
                    num += '-'
                    self.pos += 1
                while self.pos < self.length and (self.source[self.pos].isdigit() or self.source[self.pos] == '.'):
                    num += self.source[self.pos]
                    self.pos += 1
                tokens.append(Token('NUMBER', num, self.line, self.col))
                self.col += len(num)
                continue

            # 复合符号（多字符）
            two_char = ch + self.peek(1)
            if two_char in SYMBOL_MAP:
                self.pos += 2
                tokens.append(Token('SYMBOL', SYMBOL_MAP[two_char], self.line, self.col))
                self.col += 2
                continue

            # 单字符符号（& 需要特殊处理，因为 & 也可能是 && 的一部分）
            if ch == '&':
                # 前面已经处理了 &&，这里处理单独的 &
                self.pos += 1
                tokens.append(Token('SYMBOL', '&', self.line, self.col))
                self.col += 1
                continue

            if ch == '|':
                # 前面已经处理了 ||
                self.pos += 1
                self.col += 1
                continue

            # 其他单符号
            if ch in SYMBOLS_SINGLE:
                tokens.append(Token('SYMBOL', ch, self.line, self.col))
                self.pos += 1
                self.col += 1
                continue

            # 关键字或标识符
            m = re.match(r'[a-zA-Z_][a-zA-Z0-9_]*', self.source[self.pos:])
            if m:
                word = m.group(0)
                tok_type = 'KEYWORD' if word in KEYWORDS else 'IDENTIFIER'
                tokens.append(Token(tok_type, word, self.line, self.col))
                self.pos += len(word)
                self.col += len(word)
                continue

            # 忽略其他
            self.pos += 1
            self.col += 1

        tokens.append(Token('EOF', '', self.line, self.col))
        return tokens

File: php/Lexer.php
<?php
// Lexer.php - 词法分析器（支持复合符号）

class Lexer {
    private $source;
    private $length;
    private $pos;
    private $line;
    private $col;

    public function __construct($source) {
        $this->source = $source;
        $this->length = strlen($source);
        $this->pos = 0;
        $this->line = 1;
        $this->col = 1;
    }

    public function tokenize() {
        $tokens = [];
        while ($this->pos < $this->length) {
            $ch = $this->source[$this->pos];

            // 跳过空白
            if (ctype_space($ch)) {
                if ($ch === "\n") { $this->line++; $this->col = 1; } else { $this->col++; }
                $this->pos++;
                continue;
            }

            // 注释 // ...
            if ($ch === '/' && $this->peek(1) === '/') {
                $this->pos += 2;
                while ($this->pos < $this->length && $this->source[$this->pos] !== "\n") {
                    $this->pos++;
                }
                continue;
            }

            // @ 控件
            if ($ch === '@') {
                $this->pos++;
                $name = '';
                while ($this->pos < $this->length && preg_match('/[a-zA-Z0-9_:]/', $this->source[$this->pos])) {
                    $name .= $this->source[$this->pos];
                    $this->pos++;
                }
                $tokens[] = ['type' => 'CONTROL', 'value' => $name, 'line' => $this->line, 'col' => $this->col];
                $this->col += strlen($name) + 1;
                continue;
            }

            // 变量 $xxx
            if ($ch === '$') {
                $this->pos++;
                $name = '';
                while ($this->pos < $this->length && preg_match('/[a-zA-Z0-9_]/', $this->source[$this->pos])) {
                    $name .= $this->source[$this->pos];
                    $this->pos++;
                }
                $tokens[] = ['type' => 'VARIABLE', 'value' => $name, 'line' => $this->line, 'col' => $this->col];
                $this->col += strlen($name) + 1;
                continue;
            }

            // 字符串
            if ($ch === '"' || $ch === "'") {
                $quote = $ch;
                $this->pos++;
                $str = '';
                while ($this->pos < $this->length && $this->source[$this->pos] !== $quote) {
                    if ($this->source[$this->pos] === '\\') {
                        $this->pos++;
                        $str .= $this->source[$this->pos] ?? '';
                    } else {
                        $str .= $this->source[$this->pos];
                    }
                    $this->pos++;
                }
                $this->pos++; // 跳过结束引号
                $tokens[] = ['type' => 'STRING', 'value' => $str, 'line' => $this->line, 'col' => $this->col];
                $this->col += strlen($str) + 2;
                continue;
            }

            // 数字
            if (is_numeric($ch) || ($ch === '-' && $this->pos + 1 < $this->length && is_numeric($this->source[$this->pos + 1]))) {
                $num = '';
                if ($ch === '-') { $num .= '-'; $this->pos++; }
                while ($this->pos < $this->length && (is_numeric($this->source[$this->pos]) || $this->source[$this->pos] === '.')) {
                    $num .= $this->source[$this->pos];
                    $this->pos++;
                }
                $tokens[] = ['type' => 'NUMBER', 'value' => $num, 'line' => $this->line, 'col' => $this->col];
                $this->col += strlen($num);
                continue;
            }

            // ---- 复合符号处理 ----
            if ($ch === '=') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '=') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '==', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } elseif ($this->pos < $this->length && $this->source[$this->pos] === '>') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '=>', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '=', 'line' => $this->line, 'col' => $this->col];
                    $this->col++;
                }
                continue;
            }
            if ($ch === '!') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '=') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '!=', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    // 单独 ! 忽略或作为错误
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '!', 'line' => $this->line, 'col' => $this->col];
                    $this->col++;
                }
                continue;
            }
            if ($ch === '<') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '=') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '<=', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '<', 'line' => $this->line, 'col' => $this->col];
                    $this->col++;
                }
                continue;
            }
            if ($ch === '>') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '=') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '>=', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '>', 'line' => $this->line, 'col' => $this->col];
                    $this->col++;
                }
                continue;
            }
            if ($ch === '&') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '&') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '&&', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '&', 'line' => $this->line, 'col' => $this->col];
                    $this->col++;
                }
                continue;
            }
            if ($ch === '|') {
                $this->pos++;
                if ($this->pos < $this->length && $this->source[$this->pos] === '|') {
                    $this->pos++;
                    $tokens[] = ['type' => 'SYMBOL', 'value' => '||', 'line' => $this->line, 'col' => $this->col];
                    $this->col += 2;
                } else {
                    // 单竖线忽略
                    $this->col++;
                }
                continue;
            }

            // -> 符合符号（JSON 路径指示器）
            if ($ch === '-' && $this->pos + 1 < $this->length && $this->source[$this->pos + 1] === '>') {
                $this->pos += 2;
                $tokens[] = ['type' => 'SYMBOL', 'value' => '->', 'line' => $this->line, 'col' => $this->col];
                $this->col += 2;
                continue;
            }

            // 其他单个符号：{}();,:[] 
            if (strpos('{}();:,&=+-*/%[]', $ch) !== false) {
                $tokens[] = ['type' => 'SYMBOL', 'value' => $ch, 'line' => $this->line, 'col' => $this->col];
                $this->pos++;
                $this->col++;
                continue;
            }

            // 关键字或标识符
            if (preg_match('/^[a-zA-Z_][a-zA-Z0-9_]*/', substr($this->source, $this->pos), $match)) {
                $word = $match[0];
                $keywords = ['if','else','for','while','break','continue','switch','case','default','return','true','false'];
                if (in_array($word, $keywords)) {
                    $tokens[] = ['type' => 'KEYWORD', 'value' => $word, 'line' => $this->line, 'col' => $this->col];
                } else {
                    $tokens[] = ['type' => 'IDENTIFIER', 'value' => $word, 'line' => $this->line, 'col' => $this->col];
                }
                $this->pos += strlen($word);
                $this->col += strlen($word);
                continue;
            }

            // 忽略其他字符
            $this->pos++;
            $this->col++;
        }
        $tokens[] = ['type' => 'EOF', 'value' => null, 'line' => $this->line, 'col' => $this->col];
        return $tokens;
    }

    private function peek($offset) {
        $pos = $this->pos + $offset;
        return ($pos < $this->length) ? $this->source[$pos] : null;
    }
}